### PR TITLE
Bulk and batch retrieval of events.

### PR DESCRIPTION
**Pull request by erikjohnston**
This speeds up things like initial sync by batching up many requests for events and then fetching them in bulk from the database.

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -430,54 +466,240 @@ def _invalidate_get_event_cache(self, event_id):
...
+        event_list = []
+        i = 0
+        while True:
+            try:
+                with self._event_fetch_lock:
+                    event_list = self._event_fetch_list
+                    self._event_fetch_list = []
+
+                    if not event_list:
+                        if self.database_engine.single_threaded or i > 3:
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: Why "3"?

File: synapse/storage/events.py [hunk 2]
```diff
@@ -430,54 +466,240 @@ def _invalidate_get_event_cache(self, event_id):
...
+
+        events_d = defer.Deferred()
+        with self._event_fetch_lock:
+            self._event_fetch_list.append(
+                (events, events_d)
+            )
+
+            self._event_fetch_lock.notify()
+
+            if self._event_fetch_ongoing < 3:
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: Is this "3" the same as the previous "3"?

**Comment by NegativeMjark** [comment]
``_do_fetch`` seems to be reimplementing the dispatch to threadpool code in twisted.python.threadpool. How much performance gain does using a custom cross-thread dispatcher give us?

**Comment by erikjohnston** [comment]
> _do_fetch seems to be reimplementing the dispatch to threadpool code in twisted.python.threadpool. How much performance gain does using a custom cross-thread dispatcher give us?

How so? My understanding was that `twisted.python.threadpool` allowed you to dispatch a function to a threadpool, which is what `runWithConnection` is doing. Neither of which gives you primitives to do a producer/consumer model. The closest I saw was `Queue.Queue` but that doesn't allow you to drain the queue.

File: synapse/storage/events.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/storage/events.py]: The various constants here should probably be configurable.

> Does this mean we will lock up a database thread for 300ms each time we get a list of events?
Yes.

> How much does locking up the database waiting for more requests for events help performance?

It helps, though I don't have any exact numbers on me.

**Comment by NegativeMjark** [comment]
> Neither of which gives you primitives to do a producer/consumer model.

I guess my question is "How much performance boost do you get by implementing a cross-thread producer/consumer model rather than collecting the events in the main thread and using the existing cross-thread dispatch mechanism?"
My concern being whether all the extra complexity being added here can be justified.

**Comment by NegativeMjark** [comment]
A few comments along the lines of "# Waiting for 100ms for 3 times sped up initial-sync on jki.re by foo milliseconds running v0.9.bar" might make it a bit clearer to the reader why it is doing some of the things it is doing.

File: synapse/storage/events.py [hunk 3]
```diff
@@ -34,6 +35,16 @@
 logger = logging.getLogger(__name__)
 
 
+# These values are used in the `enqueus_event` and `_do_fetch` methods to
+# control how we batch/bulk fetch events from the database.
+# The values are plucked out of thing air to make initial sync run faster
+# on jki.re
+# TODO: Make these configurable.
+EVENT_QUEUE_THREADS = 3  # Max number of threads that will fetch events
+EVENT_QUEUE_ITERATIONS = 3  # No. times we block waiting for events
```
Comment (NegativeMjark) [inline synapse/storage/events.py]: s/for events/for requests for events/ ?

**Comment by NegativeMjark** [comment]
Other than wording of the comments LGTM.

**Pull request merged**